File: src/python/sp/test/extrapolate.py
#!/usr/bin/env python
# vim: set fileencoding=utf-8 :
# Laurent El Shafey <Laurent.El-Shafey@idiap.ch>
# Tue Sep 27 23:26:46 2011 +0200
#
# Copyright (C) 2011-2012 Idiap Reasearch Institute, Martigny, Switzerland
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os, sys
import unittest
import bob
import numpy

#############################################################################
# Tests blitz-based extrapolation implementation with values returned 
#############################################################################

########################## Values used for the computation ##################
eps = 1e-3
a5 = numpy.array([1,2,3,4,5], 'float64')
a14_zeros = numpy.array([0,0,0,0,1,2,3,4,5,0,0,0,0,0], 'float64')
a14_twos = numpy.array([2,2,2,2,1,2,3,4,5,2,2,2,2,2], 'float64')
a14_nearest = numpy.array([1,1,1,1,1,2,3,4,5,5,5,5,5,5], 'float64')
a14_circular = numpy.array([2,3,4,5,1,2,3,4,5,1,2,3,4,5], 'float64')
a14_mirror = numpy.array([4,3,2,1,1,2,3,4,5,5,4,3,2,1], 'float64')


A22 = numpy.array([1,2,3,4], 'float64').reshape(2,2)
A44_zeros = numpy.array([0,0,0,0,0,1,2,0,0,3,4,0,0,0,0,0], 'float64').reshape(4,4)
A44_twos = numpy.array([2,2,2,2,2,1,2,2,2,3,4,2,2,2,2,2], 'float64').reshape(4,4)
A44_nearest = numpy.array([1,1,2,2,1,1,2,2,3,3,4,4,3,3,4,4], 'float64').reshape(4,4)
A44_circular = numpy.array([4,3,4,3,2,1,2,1,4,3,4,3,2,1,2,1], 'float64').reshape(4,4)
A44_mirror = numpy.array([1,1,2,2,1,1,2,2,3,3,4,4,3,3,4,4], 'float64').reshape(4,4)
#############################################################################


def compare(v1, v2, width):
  return abs(v1-v2) <= width

def test_extrapolate_1D(res, reference, obj):
  # Tests the extrapolation
  obj.assertEqual(res.shape, reference.shape)
  for i in range(res.shape[0]):
    obj.assertTrue(compare(res[i], reference[i], eps))

def test_extrapolate_2D(res, reference, obj):
  # Tests the extrapolation
  obj.assertEqual(res.shape, reference.shape)
  for i in range(res.shape[0]):
    for j in range(res.shape[1]):
      obj.assertTrue(compare(res[i,j], reference[i,j], eps))


##################### Unit Tests ##################  
class ExtrapolationTest(unittest.TestCase):
  """Performs extrapolation product"""

##################### Convolution Tests ##################  
  def test_extrapolation_1D_zeros(self):
    b = numpy.zeros((14,), 'float64')
    bob.sp.extrapolateZero(a5,b)
    test_extrapolate_1D(b,a14_zeros,self)

  def test_extrapolation_1D_twos(self):
    b = numpy.zeros((14,), 'float64')
    bob.sp.extrapolateConstant(a5,b,2.)
    test_extrapolate_1D(b,a14_twos,self)

  def test_extrapolation_1D_nearest(self):
    b = numpy.zeros((14,), 'float64')
    bob.sp.extrapolateNearest(a5,b)
    test_extrapolate_1D(b,a14_nearest,self)

  def test_extrapolation_1D_circular(self):
    b = numpy.zeros((14,), 'float64')
    bob.sp.extrapolateCircular(a5,b)
    test_extrapolate_1D(b,a14_circular,self)

  def test_extrapolation_1D_mirror(self):
    b = numpy.zeros((14,), 'float64')
    bob.sp.extrapolateMirror(a5,b)
    test_extrapolate_1D(b,a14_mirror,self)

  def test_extrapolation_2D_zeros(self):
    B = numpy.zeros((4,4), 'float64')
    bob.sp.extrapolateZero(A22,B)
    test_extrapolate_2D(B,A44_zeros,self)

  def test_extrapolation_2D_twos(self):
    B = numpy.zeros((4,4), 'float64')
    bob.sp.extrapolateConstant(A22,B,2.)
    test_extrapolate_2D(B,A44_twos,self)

  def test_extrapolation_2D_nearest(self):
    B = numpy.zeros((4,4), 'float64')
    bob.sp.extrapolateNearest(A22,B)
    test_extrapolate_2D(B,A44_nearest,self)
  
  def test_extrapolation_2D_circular(self):
    B = numpy.zeros((4,4), 'float64')
    bob.sp.extrapolateCircular(A22,B)
    test_extrapolate_2D(B,A44_circular,self)

  def test_extrapolation_2D_mirror(self):
    B = numpy.zeros((4,4), 'float64')
    bob.sp.extrapolateMirror(A22,B)
    test_extrapolate_2D(B,A44_mirror,self)


##################### Main ##################  
if __name__ == '__main__':
  sys.argv.append('-v')
  if os.environ.has_key('BOB_PROFILE') and \
      os.environ['BOB_PROFILE'] and \
      hasattr(bob.core, 'ProfilerStart'):
    bob.core.ProfilerStart(os.environ['BOB_PROFILE'])
  os.chdir(os.path.realpath(os.path.dirname(sys.argv[0])))
  unittest.main()
  if os.environ.has_key('BOB_PROFILE') and \
      os.environ['BOB_PROFILE'] and \
      hasattr(bob.core, 'ProfilerStop'):
    bob.core.ProfilerStop()

File: src/python/sp/src/extrapolate.cc
/**
 * @file python/sp/src/extrapolate.cc
 * @date Tue Sep 27 23:26:46 2011 +0200
 * @author Laurent El Shafey <Laurent.El-Shafey@idiap.ch>
 *
 * @brief Binds extrapolation to python
 *
 * Copyright (C) 2011 Idiap Reasearch Institute, Martigny, Switzerland
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "core/python/ndarray.h"
#include "sp/extrapolate.h"

using namespace boost::python;
namespace tp = bob::python;
namespace sp = bob::sp;
namespace ca = bob::core::array;

template <typename T, int N>
static void inner_extrapolateConstant_dim_size(tp::const_ndarray a,
    tp::ndarray b, object c) {
  blitz::Array<T,N> b_ = b.bz<T,N>();
  sp::extrapolateConstant<T>(a.bz<T,N>(), b_, extract<T>(c));
}

template <typename T>
static void inner_extrapolateConstant_dim(size_t nd, tp::const_ndarray a,
    tp::ndarray b, object c) {
  switch (nd) {
    case 1: return inner_extrapolateConstant_dim_size<T,1>(a,b,c);
    case 2: return inner_extrapolateConstant_dim_size<T,2>(a,b,c);
    default: PYTHON_ERROR(TypeError, "constant extrapolation not supported for array with " SIZE_T_FMT " dimensions", nd);
  }
}

static void extrapolateConstant(tp::const_ndarray a, tp::ndarray b, object c) {
  const ca::typeinfo& info = a.type();
  switch (info.dtype) {
    case ca::t_bool: 
      return inner_extrapolateConstant_dim<bool>(info.nd, a,b,c);
    case ca::t_int8: 
      return inner_extrapolateConstant_dim<int8_t>(info.nd, a,b,c);
    case ca::t_int16: 
      return inner_extrapolateConstant_dim<int16_t>(info.nd, a,b,c);
    case ca::t_int32: 
      return inner_extrapolateConstant_dim<int32_t>(info.nd, a,b,c);
    case ca::t_int64: 
      return inner_extrapolateConstant_dim<int64_t>(info.nd, a,b,c);
    case ca::t_uint8: 
      return inner_extrapolateConstant_dim<uint8_t>(info.nd, a,b,c);
    case ca::t_uint16:
      return inner_extrapolateConstant_dim<uint16_t>(info.nd, a,b,c);
    case ca::t_uint32: 
      return inner_extrapolateConstant_dim<uint32_t>(info.nd, a,b,c);
    case ca::t_uint64: 
      return inner_extrapolateConstant_dim<uint64_t>(info.nd, a,b,c);
    case ca::t_float32:
      return inner_extrapolateConstant_dim<float>(info.nd, a,b,c);
    case ca::t_float64: 
      return inner_extrapolateConstant_dim<double>(info.nd, a,b,c);
    case ca::t_complex64: 
      return inner_extrapolateConstant_dim<std::complex<float> >(info.nd, a,b,c);
    case ca::t_complex128: 
      return inner_extrapolateConstant_dim<std::complex<double> >(info.nd, a,b,c);
    default: PYTHON_ERROR(TypeError, "constant extrapolation not supported for array with type '%s'", info.str().c_str());
  }
}

template <typename T, int N>
static void inner_extrapolateZero_dim_size(tp::const_ndarray a,
    tp::ndarray b) {
  blitz::Array<T,N> b_ = b.bz<T,N>();
  sp::extrapolateZero<T>(a.bz<T,N>(), b_);
}

template <typename T>
static void inner_extrapolateZero_dim(size_t nd, tp::const_ndarray a,
    tp::ndarray b) {
  switch (nd) {
    case 1: return inner_extrapolateZero_dim_size<T,1>(a,b);
    case 2: return inner_extrapolateZero_dim_size<T,2>(a,b);
    default: PYTHON_ERROR(TypeError, "zero extrapolation not supported for array with " SIZE_T_FMT " dimensions", nd);
  }
}

static void extrapolateZero(tp::const_ndarray a, tp::ndarray b) {
  const ca::typeinfo& info = a.type();
  switch (info.dtype) {
    case ca::t_bool: 
      return inner_extrapolateZero_dim<bool>(info.nd, a,b);
    case ca::t_int8: 
      return inner_extrapolateZero_dim<int8_t>(info.nd, a,b);
    case ca::t_int16: 
      return inner_extrapolateZero_dim<int16_t>(info.nd, a,b);
    case ca::t_int32: 
      return inner_extrapolateZero_dim<int32_t>(info.nd, a,b);
    case ca::t_int64: 
      return inner_extrapolateZero_dim<int64_t>(info.nd, a,b);
    case ca::t_uint8: 
      return inner_extrapolateZero_dim<uint8_t>(info.nd, a,b);
    case ca::t_uint16:
      return inner_extrapolateZero_dim<uint16_t>(info.nd, a,b);
    case ca::t_uint32: 
      return inner_extrapolateZero_dim<uint32_t>(info.nd, a,b);
    case ca::t_uint64: 
      return inner_extrapolateZero_dim<uint64_t>(info.nd, a,b);
    case ca::t_float32:
      return inner_extrapolateZero_dim<float>(info.nd, a,b);
    case ca::t_float64: 
      return inner_extrapolateZero_dim<double>(info.nd, a,b);
    case ca::t_complex64: 
      return inner_extrapolateZero_dim<std::complex<float> >(info.nd, a,b);
    case ca::t_complex128: 
      return inner_extrapolateZero_dim<std::complex<double> >(info.nd, a,b);
    default: PYTHON_ERROR(TypeError, "zero extrapolation not supported for array with type '%s'", info.str().c_str());
  }
}

template <typename T, int N>
static void inner_extrapolateNearest_dim_size(tp::const_ndarray a,
    tp::ndarray b) {
  blitz::Array<T,N> b_ = b.bz<T,N>();
  sp::extrapolateNearest<T>(a.bz<T,N>(), b_);
}

template <typename T>
static void inner_extrapolateNearest_dim(size_t nd, tp::const_ndarray a,
    tp::ndarray b) {
  switch (nd) {
    case 1: return inner_extrapolateNearest_dim_size<T,1>(a,b);
    case 2: return inner_extrapolateNearest_dim_size<T,2>(a,b);
    default: PYTHON_ERROR(TypeError, "nearest extrapolation not supported for array with " SIZE_T_FMT " dimensions", nd);
  }
}

static void extrapolateNearest(tp::const_ndarray a, tp::ndarray b) {
  const ca::typeinfo& info = a.type();
  switch (info.dtype) {
    case ca::t_bool: 
      return inner_extrapolateNearest_dim<bool>(info.nd, a,b);
    case ca::t_int8: 
      return inner_extrapolateNearest_dim<int8_t>(info.nd, a,b);
    case ca::t_int16: 
      return inner_extrapolateNearest_dim<int16_t>(info.nd, a,b);
    case ca::t_int32: 
      return inner_extrapolateNearest_dim<int32_t>(info.nd, a,b);
    case ca::t_int64: 
      return inner_extrapolateNearest_dim<int64_t>(info.nd, a,b);
    case ca::t_uint8: 
      return inner_extrapolateNearest_dim<uint8_t>(info.nd, a,b);
    case ca::t_uint16:
      return inner_extrapolateNearest_dim<uint16_t>(info.nd, a,b);
    case ca::t_uint32: 
      return inner_extrapolateNearest_dim<uint32_t>(info.nd, a,b);
    case ca::t_uint64: 
      return inner_extrapolateNearest_dim<uint64_t>(info.nd, a,b);
    case ca::t_float32:
      return inner_extrapolateNearest_dim<float>(info.nd, a,b);
    case ca::t_float64: 
      return inner_extrapolateNearest_dim<double>(info.nd, a,b);
    case ca::t_complex64: 
      return inner_extrapolateNearest_dim<std::complex<float> >(info.nd, a,b);
    case ca::t_complex128: 
      return inner_extrapolateNearest_dim<std::complex<double> >(info.nd, a,b);
    default: PYTHON_ERROR(TypeError, "nearest extrapolation not supported for array with type '%s'", info.str().c_str());
  }
}

template <typename T, int N>
static void inner_extrapolateCircular_dim_size(tp::const_ndarray a,
    tp::ndarray b) {
  blitz::Array<T,N> b_ = b.bz<T,N>();
  sp::extrapolateCircular<T>(a.bz<T,N>(), b_);
}

template <typename T>
static void inner_extrapolateCircular_dim(size_t nd, tp::const_ndarray a,
    tp::ndarray b) {
  switch (nd) {
    case 1: return inner_extrapolateCircular_dim_size<T,1>(a,b);
    case 2: return inner_extrapolateCircular_dim_size<T,2>(a,b);
    default: PYTHON_ERROR(TypeError, "circular extrapolation not supported for array with " SIZE_T_FMT " dimensions", nd);
  }
}

static void extrapolateCircular(tp::const_ndarray a, tp::ndarray b) {
  const ca::typeinfo& info = a.type();
  switch (info.dtype) {
    case ca::t_bool: 
      return inner_extrapolateCircular_dim<bool>(info.nd, a,b);
    case ca::t_int8: 
      return inner_extrapolateCircular_dim<int8_t>(info.nd, a,b);
    case ca::t_int16: 
      return inner_extrapolateCircular_dim<int16_t>(info.nd, a,b);
    case ca::t_int32: 
      return inner_extrapolateCircular_dim<int32_t>(info.nd, a,b);
    case ca::t_int64: 
      return inner_extrapolateCircular_dim<int64_t>(info.nd, a,b);
    case ca::t_uint8: 
      return inner_extrapolateCircular_dim<uint8_t>(info.nd, a,b);
    case ca::t_uint16:
      return inner_extrapolateCircular_dim<uint16_t>(info.nd, a,b);
    case ca::t_uint32: 
      return inner_extrapolateCircular_dim<uint32_t>(info.nd, a,b);
    case ca::t_uint64: 
      return inner_extrapolateCircular_dim<uint64_t>(info.nd, a,b);
    case ca::t_float32:
      return inner_extrapolateCircular_dim<float>(info.nd, a,b);
    case ca::t_float64: 
      return inner_extrapolateCircular_dim<double>(info.nd, a,b);
    case ca::t_complex64: 
      return inner_extrapolateCircular_dim<std::complex<float> >(info.nd, a,b);
    case ca::t_complex128: 
      return inner_extrapolateCircular_dim<std::complex<double> >(info.nd, a,b);
    default: PYTHON_ERROR(TypeError, "circular extrapolation not supported for array with type '%s'", info.str().c_str());
  }
}

template <typename T, int N>
static void inner_extrapolateMirror_dim_size(tp::const_ndarray a,
    tp::ndarray b) {
  blitz::Array<T,N> b_ = b.bz<T,N>();
  sp::extrapolateMirror<T>(a.bz<T,N>(), b_);
}

template <typename T>
static void inner_extrapolateMirror_dim(size_t nd, tp::const_ndarray a,
    tp::ndarray b) {
  switch (nd) {
    case 1: return inner_extrapolateMirror_dim_size<T,1>(a,b);
    case 2: return inner_extrapolateMirror_dim_size<T,2>(a,b);
    default: PYTHON_ERROR(TypeError, "mirror extrapolation not supported for array with " SIZE_T_FMT " dimensions", nd);
  }
}

static void extrapolateMirror(tp::const_ndarray a, tp::ndarray b) {
  const ca::typeinfo& info = a.type();
  switch (info.dtype) {
    case ca::t_bool: 
      return inner_extrapolateMirror_dim<bool>(info.nd, a,b);
    case ca::t_int8: 
      return inner_extrapolateMirror_dim<int8_t>(info.nd, a,b);
    case ca::t_int16: 
      return inner_extrapolateMirror_dim<int16_t>(info.nd, a,b);
    case ca::t_int32: 
      return inner_extrapolateMirror_dim<int32_t>(info.nd, a,b);
    case ca::t_int64: 
      return inner_extrapolateMirror_dim<int64_t>(info.nd, a,b);
    case ca::t_uint8: 
      return inner_extrapolateMirror_dim<uint8_t>(info.nd, a,b);
    case ca::t_uint16:
      return inner_extrapolateMirror_dim<uint16_t>(info.nd, a,b);
    case ca::t_uint32: 
      return inner_extrapolateMirror_dim<uint32_t>(info.nd, a,b);
    case ca::t_uint64: 
      return inner_extrapolateMirror_dim<uint64_t>(info.nd, a,b);
    case ca::t_float32:
      return inner_extrapolateMirror_dim<float>(info.nd, a,b);
    case ca::t_float64: 
      return inner_extrapolateMirror_dim<double>(info.nd, a,b);
    case ca::t_complex64: 
      return inner_extrapolateMirror_dim<std::complex<float> >(info.nd, a,b);
    case ca::t_complex128: 
      return inner_extrapolateMirror_dim<std::complex<double> >(info.nd, a,b);
    default: PYTHON_ERROR(TypeError, "mirror extrapolation not supported for array with type '%s'", info.str().c_str());
  }
}

void bind_sp_extrapolate() {

  def("extrapolateConstant", &extrapolateConstant, (arg("src"), arg("dst")), "Extrapolates the values of a 1D array with a constant, given a 1 or 2D input array.");

  def("extrapolateZero", &extrapolateZero, (arg("src"), arg("dst")), "Extrapolates the values of a 1D array with zeros, given a 1 or 2D input array.");

  def("extrapolateNearest", &extrapolateNearest, (arg("src"), arg("dst")), "Extrapolates the values of a 1D array with nearest values, given a 1 or 2D input array.");

  def("extrapolateCircular", &extrapolateCircular, (arg("src"), arg("dst")), "Extrapolates the values of a 1D array using circular extrapolation, given a 1 or 2D input array.");

  def("extrapolateMirror", &extrapolateMirror, (arg("src"), arg("dst")), "Extrapolates the values of a 1D array using mirroring, given a 1 or 2D input array.");

}
